extensions-contrib/moving-average-query/src/main/java/org/apache/druid/query/movingaverage/MovingAverageIterable.java

```diff
@@ -0,0 +1,316 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * {@link MovingAverageIterable} iterates over days {@link RowBucket}, producing rows for each dimension combination,
+ * filling in missing entries with "empty" rows so that the averaging buckets have enough data to operate on.
+ * It then computes the moving average on the buckets and returns the row.
+ * See {@link MovingAverageIterator#computeMovingAverage(Map, Row, boolean)} for more details.
+ */
+public class MovingAverageIterable implements Iterable<Row>
+{
+
+  private final Sequence<RowBucket> seq;
+  private final Collection<DimensionSpec> dims;
+  private final Collection<AveragerFactory<?, ?>> factories;
+  private final Map<String, PostAggregator> postAggMap;
+  private final Map<String, AggregatorFactory> aggMap;
+  private final Map<String, Object> fakeEvents;
+
+  public MovingAverageIterable(
+      Sequence<RowBucket> buckets,
+      Collection<DimensionSpec> dims,
+      Collection<AveragerFactory<?, ?>> factories,
+      List<PostAggregator> postAggList,
+      List<AggregatorFactory> aggList
+  )
+  {
+    this.dims = dims;
+    this.factories = factories;
+    this.seq = buckets;
+
+    postAggMap = postAggList.stream().collect(Collectors.toMap(postAgg -> postAgg.getName(), postAgg -> postAgg));
+    aggMap = aggList.stream().collect(Collectors.toMap(agg -> agg.getName(), agg -> agg));
+    fakeEvents = generateFakeEventsFromAggregators(aggMap, postAggMap);
+  }
+
+  // Build a list of dummy events from Aggregators/PostAggregators to be used by Iterator to build fake rows.
+  // These fake rows will be used by computeMovingAverage() in skip=true mode.
+  // See fakeEventsCopy in internalNext() and computeMovingAverage() documentation.
+  private Map<String, Object> generateFakeEventsFromAggregators(Map<String, AggregatorFactory> aggMap,
+                                                                Map<String, PostAggregator> postAggMap)
+  {
+    Map<String, Object> fakeEvents = new LinkedHashMap<>();
+    aggMap.values().forEach(agg -> {
+      Aggregator aggFactorized = agg.factorize(getEmptyColumnSelectorFactory());
+      fakeEvents.put(agg.getName(), aggFactorized.get());
+    });
+    postAggMap.values().forEach(postAgg -> fakeEvents.put(postAgg.getName(), postAgg.compute(fakeEvents)));
+    return fakeEvents;
+  }
+
+  @Nonnull
+  private ColumnSelectorFactory getEmptyColumnSelectorFactory()
+  {
+    return new ColumnSelectorFactory()
+    {
+      @Override
+      public DimensionSelector makeDimensionSelector(DimensionSpec dimensionSpec)
+      {
+        // Generating empty records while aggregating on Filtered aggregators requires a dimension selector
+        // for initialization.  This dimension selector is not actually used for generating values
+        return DimensionSelector.constant(null);
+      }
+
+      @Override
+      public ColumnValueSelector makeColumnValueSelector(String s)
+      {
+        return null;
```

Comment:
I guess this should be `NilColumnValueSelector`. Otherwise, this will throw NPE if any aggregator needs `ColumnVauleSelector` like in `HyperUniquesAggregatorFactory`.

Reply:
Fixed.